python/ray/dag/tests/experimental/test_mocked_nccl_dag.py

```diff
@@ -0,0 +1,385 @@
+        result = output_channel.begin_read()
+        assert result == spec
+        output_channel.end_read()
+
+    ray.kill(barrier1)
+    ray.kill(barrier2)
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize(
+    "ray_start_cluster",
+    [
+        {
+            "num_cpus": 2,
+            "num_gpus": 2,
+            "num_nodes": 1,
+        }
+    ],
+    indirect=True,
+)
+def test_p2p_static_non_tensor_data_error(ray_start_cluster):
+    """
+    Test that when static_non_tensor_data=True, an error is thrown when a
+    different number of tensors is found.
+    """
+    # Barrier name should be barrier-{sender rank}-{receiver rank}.
+    # Create a barrier in both directions because we don't know which rank will
+    # get assigned to sender and receiver.
+    barrier1 = Barrier.options(name="barrier-0-1").remote()  # noqa
+    barrier2 = Barrier.options(name="barrier-1-0").remote()  # noqa
+
+    sender = MockedWorker.remote()
+    receiver = MockedWorker.remote()
+
+    ray.get([sender.start_mock.remote(), receiver.start_mock.remote()])
+
+    # Test torch.Tensor sent between actors.
+    with InputNode() as inp:
+        dag = sender.send_dict.bind(inp)
+        dag = dag.with_type_hint(
+            TorchTensorType(transport="nccl", static_non_tensor_data=True)
+        )
+        dag = receiver.recv_dict.bind(dag)
+
+    compiled_dag = dag.experimental_compile()
+
+    shape = 10
+    dtype = torch.float16
+    # Test sending a dictionary that always contains keys inserted in the same
+    # order, but with different-sized and different-valued tensors.
+    # same keys.
+    for i in range(3):
+        tensor_shape = (shape * (i + 1),)
+        spec = [
+            ("a", i, tensor_shape, dtype),
+            ("b", i, tensor_shape, dtype),
+            ("c", i, tensor_shape, dtype),
+        ]
+        output_channel = compiled_dag.execute(spec)
+        # TODO(swang): Replace with fake ObjectRef.
+        result = output_channel.begin_read()
+        assert result == spec
+        output_channel.end_read()
+
+    # Test sending a dictionary with a different number of entries.
+    spec = [
+        ("a", 1, shape, dtype),
+        ("b", 1, shape, dtype),
+    ]
+    output_channel = compiled_dag.execute(spec)
+    with pytest.raises(OSError, match="Channel closed"):
+        result = output_channel.begin_read()
+
+    # Sending correct shape still errors because the DAG has already been torn
+    # down after the previous error.
+    spec.append(("c", 1, shape, dtype))
+    with pytest.raises(OSError, match="Channel closed"):
+        output_channel = compiled_dag.execute(spec)
+
+    # TODO(swang): Ideally we would also check that sending a different value,
+    # e.g., changing the keys of the dictionary or sending a list instead of a
+    # dictionary, would also fail, but currently we don't have a reliable way
+    # to do this. Checking the serialized output produces many false positives.
```

Comment:
Can we check this in the `write()` function after calling `reset_out_of_band_tensors()`? Maybe we could zero out all of the tensor references and then take the checksum.

Reply:
Sorry, I didn't understand what you're suggesting we check?